### ISSUE TITLE
Use more specific data type when some pgv-rules are present

### ISSUE THREAD

**Issue by gontard** [scalapb/scalapb-validate#38]
We are using scalapb, scalapb-validate and [cats](https://typelevel.org/cats/), we facilitate the manipulation of the generated classes, we would like to take advantage of both 

For example in this case, we would like to expose a the items as a [`cats.NonEmptyList`](https://typelevel.org/cats/datatypes/nel.html)

```
  repeated string items  = 1 [(validate.rules).repeated.min_items = 1];
```
and here a `cats.NonEmptySet`
```
  repeated string items  = 1 [(validate.rules).repeated = { min_items: 1, unique: true }];
```

Since the validation is made on an instantiated object, the use of scalapb customization `(scalapb.field).collection_type` seems unadapted.
May be we could generate somehow a method that return the appropriate specific data type:
```
case class Message(items: Seq[String]) {
    def itemsNel: NonEmptyList[String] = NonEmptyList.fromListUnsafe(items.toList)
}
```

Is there a way to do that?

**Comment by thesamet** [comment]
It looks like we're looking at two options:
1. Have the constructor type automatically change to NEL whenever min_items is 1 and fail parsing if the rule is violated.
2. Automatically generate `xNel` methods for each repeated field `x` that has min_items set to 1.

#2 is definitely easier - since it doesn't interfere with the parsing logic. Do both options work equally well for you?

**Comment by gontard** [comment]
@thesamet Thanks for summing that up.
This idea is quite new and i don't have yet enough perspective _but_ right now the first solution is my favourite as it enforces these constraints at build time.

Since i opened the issue, i have encountered the same pattern not only for `cats.NonEmptyList` or `cats.NonEmptySet` but also for some my own datatype (for example `Email`).

**Comment by gontard** [comment]
@thesamet what would be nice is to somehow be able to map a set of PGV rules to a scala type:

```scala
def typeForRule(rules: io.envoyproxy.pgv.validate.FieldRules): Any {
    rules match {
        ....  => NonEmptySet
        ....  => Email
    }
}
```

**Comment by thesamet** [comment]
This proposal makes sense, and I think this can make a nice addition to this library. Currently, there is no way for one plugin to change the types emitted by ScalaPB. We would have to come up with a new capability in core ScalaPB that would enable this. This appears to be high effort, and only applicable to a specific user group (scalapb-validate && Cats). It would be great if this work can be sponsored by one of the companies who use it, or someone putting together the necessary design and PRs for discussion.

**Comment by bjaglin** [comment]
For the record, I am happy to announce that [Teads](https://medium.com/teads-engineering) will be sponsoring this work!

**Comment by bjaglin** [comment]
Most of the PGV rules can "manually" be encoded/enforced by types by annotating fields with a `scalapb` option (provided that scalapb-validate is disabled, as it would fail on these custom types as @gontard noted above).

1. Collections - `(validate.rules).repeated` & `(validate.rules).map`
   * `(scalapb.field).collection_type`
   * `(scalapb.field).map_type`
1. Scalar - `(validate.rules).<scalar>`, `(validate.rules).enum`, `(validate.rules).map.keys` & `(validate.rules).map.values`
   * `(scalapb.field).type`
   * `(scalapb.field).key_type`
   * `(scalapb.field). value_type` 
1. Required - `(validate.rules).message.required`
   * ? (no way to skip the `Option` boxing)

This issue was originally targeted at (1), but (2) was introduced in the discussion.

In our own use-case, (1) is the most interesting, (2) could be useful, and (3) is currently irrelevant as we use [proto2](https://github.com/scalapb/scalapb-validate/issues/21) (but having that feature would make it possible for us to rather easily switch to proto3 as the ability to declare fields `required` and avoid option unwrapping ceremony is what is holding us back).

**Comment by bjaglin** [comment]
scalapb-validate currently produce separate objects with a `validate()` method that the client must call explicitly. If a new `scalapb-validate-cats` module is provided with NEL/NES presets (for example), ScalaPB stubs themselves would be impacted, raising exceptions at object creation/unmarshalling time without any client action. However, the types might not capture all PGV constraints, so validators should still be supported with these custom stubs I suppose. Does it make sense in that case to automatically call the validators somehow to avoid partial validation?

**Comment by thesamet** [comment]
> However, the types might not capture all PGV constraints, so scalapb-validate validator objects should still be supported with these custom stubs I suppose. Does it make sense in that case to automatically call the validators somehow to avoid partial validation?
If I understand correctly the preference is to have the case class constructor call pgv-validate so it is impossible to instantiate directly or through parsing partially invalid objects?

**Comment by bjaglin** [comment]
> If I understand correctly the preference is to have the case class constructor call pgv-validate so it is impossible to instantiate directly or through parsing partially invalid objects?

That's correct. As a matter of fact, in our `akka-grpc` context that's what we do to enforce the rules on incoming request payloads, with an internal (fragile) subclass of `scalapb.compiler.ProtobufGenerator` adding a `validator` field and mixing-in
```scala
trait ScalaPBValidator[ScalaPB] {
  self: ScalaPB =>
  def validator: Validator[ScalaPB]

  validator.validate(self) match {
    case Success => ()
    case Failure(violations) =>
      val description = violations.map(_.getMessage).mkString("\n")
      throw new GrpcServiceException(INVALID_ARGUMENT.withDescription(description))
  }
}
```

I guess it will depend on the implementation, but what we could loose if we encode a repeated field with NEL would be a clear exception with the violated validations (if the NEL constructor is called first). That's why I brought up the subject of partial validation, but I haven't given it much thoughst yet.

**Comment by thesamet** [comment]
Do we always want to have `Set`/`NonEmptySet` when `unique: true` is given? When converting to `Set` we lose the original order.

**Comment by bjaglin** [comment]
> Do we always want to have `Set`/`NonEmptySet` when `unique: true` is given? When converting to `Set` we lose the original order.

Good point.

At the moment we don't have a case where we need "ordered list without duplicates" semantics, but that's what it really is.

Which brings the question of opting-in for NES usage i guess. This might be out of scope for this, and I don't know how scalapb-validate-cats could declare options (under it's own registered extension?), but controlling the behavior via an option would be useful. It doesn't change validation so it doesn't belong in PGV, but it does document the contract for non-scalapb users (something important for us in our gRPC use case), so it's not strictly an implementation detail.

**Comment by gontard** [comment]
> Do we always want to have `Set`/`NonEmptySet` when `unique: true` is given? When converting to `Set` we lose the original order.

May be we could also rely on a `LinkedHashSet` to preserve the order.

**Comment by thesamet** [comment]
I've just cut a preview release of scalapb-validate with this functionality. Check out the [docs](https://scalapb.github.io/docs/validation/#rule-based-type-customizaton) and the [example project](https://github.com/scalapb/scalapb-validate-demo).

**Comment by piyo7** [comment]
I tried a wonderful preview release using preprocessor. How about postprocessor?
My proto3 files heavily use `(validate.rules).message.required = true`.  After required validations, unboxed types are desirable.

For the following proto3 messages:

```protobuf
syntax = "proto3";

package mypackage;

import "validate/validate.proto";

message A {
  string s = 1;
}

message B {
  A a = 1 [(validate.rules).message.required = true];
}

message C {
  B b = 1 [(validate.rules).message.required = true];
}
```

I currently make helper classes such as the following.

```scala
package mypackage.sample.validated

import mypackage.{sample => pre}

case class B(a: A)
object B {
  def postprocess(value: pre.B): B = B(value.a.get)
}

case class C(b: B)
object C {
  def postprocess(value: pre.C): C = C(B.postprocess(value.b.get))
}
```

```scala
import mypackage.sample._
import mypackage.sample.validated
import scalapb.validate.{Validator, Failure, Success}

object Validator2 {
  def postprocess[T: Validator, U](value: T)(implicit postprocess: Postprocess[T, U]): Either[List[ValidationException], U] =
    Validator[T].validate(value) match {
      case Failure(violations) => Left(violations)
      case Success => Right(postprocess(value))
    }

  trait Postprocess[T, U] {
    def apply(a: T): U
  }
  implicit val b: Postprocess[B, validated.B] = validated.B.postprocess
  implicit val c: Postprocess[C, validated.C] = validated.C.postprocess
}
```

```scala
import mypackage.sample._
import mypackage.sample.validated

object Test {
  val c1 = C(Some(B(Some(A("s")))))
  Validator2.postprocess(c1) // == Right(validated.C(validated.B(A("s"))))

  val c2 = C(None)
  Validator2.postprocess(c2) // == Left(...)
}
```

I would be delighted if these classes could be generated by scalapb-validate.

The postprocessor may avoid the following problem. After an exact uniqueness validation, a Set or a NonEmptySet will start to be use instead of list.

> Therefore, when you use a Set or a NonEmptySet instead of list, duplicate elements are deduped during parsing time, when ScalaPB adds them to the collections' builder. As a result, uniqueness validation will always succeed, even when the original message had duplicate elements.

**Comment by thesamet** [comment]
@piyo7 thanks for checking out the preview. Getting PGV required fields unboxed is definitely a use case I have in mind. I am thinking of a solution that would look like this:

```
// doesn't work yet!
option (scalapb.validate.package) = {
    field_transformations: [
        {
            when: {message: {required: true}}
            set: {
              required: true
            }
        }
    ]
};
```

The would tell ScalaPB to set `required: true` (a new ScalaPB option) whenever `message.required` is true. That option would work like `no_box` - not wrap the value in `Option`, but in addition parsing from binary/json/text would throw an exception if no value is explicitly set. Would that work?

**Comment by piyo7** [comment]
Thank you for replying. I also tried settings like this suggestion with the preview. That sounds good.
Honestly, I would like to avoid exceptions, which are generally expensive in JVM. My microservice will receive requests form many clients that can violate rules, because they may be implemented by TypeScript or C# not supported by protoc-gen-validate.  For performance, my helper class returns an Either instead of throwing an exception. 
I understand that postprocess requires to generate different message classes in scalapb-validate, and there are other overheads of type conversions like boxing and unboxing.
If this suggestion is implemented, I will firstly use it. When my microservice encounter performance problems, I will manually implement validated message classes.

**Comment by thesamet** [comment]
We can have a version of `parseFrom` that returns an `Either` instead of throwing an exception, it's something I wanted to have for a while.

**Comment by piyo7** [comment]
Thanks. `required: true` and `parseFrom` are exactly what I want.

**Comment by thesamet** [comment]
Released 0.2.0-preview5 (needs ScalaPB 0.10.0-preview5), which lets you unbox optional message fields, and throw an `InvalidProtocolBufferException` if the field is missing when parsing. It's not `ValidationException` since the validation performed by ScalaPB during parsing.

I am also experimenting with version of ScalaPB that relies on `Either` internally instead of throwing, but I am getting  regressions in performance benchmarks due to the need to wrap everything in `Right`. I'll get back to no-throwing implementation in a different issue.

**Comment by bjaglin** [comment]
Some feedback regarding the [mini-DSL](https://github.com/scalapb/ScalaPB/issues/1007#issuecomment-751610600):

https://github.com/scalapb/ScalaPB/blob/master/docs/src/main/markdown/validation.md
> The order this happens is subject to change and should not be assumed.

We don't have any concrete issue/problem with that at the moment, but I wonder why you don't want to document/commit an order? YAGNI? I guess the recommendation is to write tests if we do use features subject to ordering and report back in case we end up with something not achievable down the road.

Here are some thoughts/questions about the implementation
1. Curious to see how it was working with `min_items: 1; unique: true` (matching both `NES` & `Set`), I checked the implementation. I saw that field transformations are used internally by the [built-in rules](https://github.com/scalapb/scalapb-validate/blob/f5c0028f89a87d6a4fbb411abb4d2705705c1c34/code-gen/src/main/scala/scalapb/validate/compiler/ValidatePreprocessor.scala#L163-L190), and that there is an implicit, adhoc order there. I might be overloooking but actually, it looks that they are even mutually exclusive? If so, should they be within a oneof?
2. Even though we don't have the use case now, there might be times where we want to override the behavior of the built-in rules, for example for https://github.com/scalapb/scalapb-validate/issues/38#issuecomment-749956687. I haven't tested it, but are user transformations applied last?

**Comment by bjaglin** [comment]
Looking back at https://github.com/scalapb/scalapb-validate/issues/38#issuecomment-746544279, it seems that only (4) remains unimplemented in 0.2.0-preview5 (but as I mentioned, it came as an afterthought). Implementation would mostly consists of a new ScalaPB option on `oneof` to deactivate generation of the `Empty` case objects, but is that a contribution/change you would be open to merge @thesamet ? If so, I can follow-up in a separate ticket for clarity.

**Comment by thesamet** [comment]
Thanks for the feedback. Right now, `NES` and `Set` are mutually exclusive, but I think we can just always apply `NES` after `Set` since it just overwrites the output. Let me think mre on documenting the order transformations are applied. I was mostly concerned not having enough user input to get it right.

Sure, let's discuss non-empty oneofs on a separate ticket.

**Comment by thesamet** [comment]
I have a pretty exciting update available to test in `preview6`. It is now possible to extract values from the rules and reference them in the applied options. This enables, among other things, to use [refined](https://github.com/fthomas/refined) to achieve compile time validation of PGV rules.

Here are a few examples of how to use it:

```
option (scalapb.validate.package) = {
  field_transformations : [ {
    when : {int32 : {gt : 1}}  // <-- 1 can be replaced with any number
    set : {type : "Int Refined Greater[$(int32.gt)]"}
    match_type : PRESENCE
  } ]

message Test {
  int32 gt_test = 1 [ (.validate.rules).int32 = {gt : 5} ];  // transformed to: Int Refined Greater[5]
}
```

We introduce a new match_type, `PRESENCE` which tells our preprocessor to only compare the presence of the fields in the when pattern, not their specific value. In the `set` part, we can refer to any field in the FieldRule by giving a path like `$(int32.gt)`.  This substitution works for all match_types.

There are more examples in [this test file](https://github.com/scalapb/scalapb-validate/blob/0.2.x-preview/e2e/src/main/protobuf/transforms/refined/refined.proto#L18-L43). Our test suite runs on both Scala 2.12 and Scala 2.13 so it relies on shapeless for literal types.

In order for all of this to work, we need a typemapper that knows to convert to/from refined types for anything that must be converted at runtime - [see here](https://github.com/scalapb/scalapb-validate/blob/0.2.x-preview/e2e/src/main/scala/scalapb/transforms/refined/package.scala).

Finally, [new tests](https://github.com/scalapb/scalapb-validate/blob/0.2.x-preview/e2e/src/test/scala/scalapb/validate/transforms/refined/RefinedSpec.scala) were added to assert the types are deployed correctly.

**Comment by bjaglin** [comment]
A last question before calling it feature-complete on our side: in our target project, all our protos are in the same package, spread across different sibling sbt modules, so we can have per-module package-level options to incrementally enable/use built-in and custom field transformations. There is a concern that the per-module option is not granular enough for big modules to transition smoothly, so I just looked at our options to enable transformations on a per-file level.

I figured out we could declare field transformations at the package level and opt-in for the preprocessor at the file level: https://github.com/bjaglin/scalapb-validate-demo/commit/774b1b28f199444e4d2ffb90ffade1f3ef5c0071.

However, my initial goal was to opt-out at the file level (since I assume it will be easy to migrate 90% of the files), but the 2 options I explored lead to a dead-end:
1. The `mergeFrom` semantics used when combining package-level and file-level options in ScalaPB options do not allow to "remove" a preprocessor defined in an outer scope.
1. scalapb-validate options (the field transformation) can only be defined at the package-level, and even if there were also at the file-level, we would probably get "append" semantics [like we rightly do for parent packages](https://github.com/scalapb/ScalaPB/commit/7ca6de3b06f54741ee0024e871758aa94d6c8316), a bit like in (1)

I guess we can live with an opt-in rather than an opt-out, but any idea or API tweak to allow opt-out is welcome!

**Comment by thesamet** [comment]
> I guess we can live with an opt-in rather than an opt-out, but any idea or API tweak to allow opt-out is welcome!

In ScalaPB 0.10.10-preview7, opting out files from preprocessing can be achieved by adding a `"-$name"`  to the preprocessors list at the file level. This change does not require upgrading scalapb-validate (verified to work with preview6). Example in 1d46cc83288d079bd1dc26e9b4ad7b6686e6db26.

**Comment by gontard** [comment]
👍  Thanks for this new option.
May be i did something wrong but when i tried, i got this exception:
```
--jvm_2_out: scalapb.compiler.GeneratorException: commands/GrantRolesForAccount.proto: Invalid preprocessor name: '-scalapb-validate-preprocessor'
	at scalapb.compiler.FileOptionsCache$.$anonfun$buildCache$16(FileOptionsCache.scala:110)
```

**Comment by thesamet** [comment]
@gontard based on the line number in the exception I think you're not on ScalaPB 0.10.10-preview7 - can you double check?

**Comment by gontard** [comment]
@thesamet indeed when i run `protocGenerate` it looks like the `0.10.10-preview6` is used. What is weird is that i depend on `0.10.10-preview7`.

```
reload plugins
evicted

[warn] 	* com.thesamet.scalapb:compilerplugin_2.12:0.10.10-preview7 is selected over {0.10.8, 0.10.8, 0.10.10-preview6, 0.9.1}
[warn] 	    +- root:root:0.1.0-SNAPSHOT (sbtVersion=1.0, scalaVersion=2.12) (depends on 0.10.10-preview7)
[warn] 	    +- tv.teads:teads-build-plugin:9.1.0 (sbtVersion=1.0, scalaVersion=2.12) (depends on 0.9.1)
[warn] 	    +- com.thesamet.scalapb:scalapb-validate-codegen_2.12:0.2.0-preview6 (depends on 0.10.10-preview6)
[warn] 	    +- com.lightbend.akka.grpc:sbt-akka-grpc:1.0.2 (sbtVersion=1.0, scalaVersion=2.12) (depends on 0.10.8)
[warn] 	    +- com.lightbend.akka.grpc:akka-grpc-codegen_2.12:1.0.2 (depends on 0.10.8)
```

I suppose that the `sbt-protoc` downloads the previous version (`0.10.10-preview7`) when the `scalapb-validate` generator is run. 

I reproduced the issue on [scalapb-validate-demo](https://github.com/scalapb/scalapb-validate-demo/commit/7622a9b35f9dbaa4f13e2956a91743c9f2c122f3)

**Comment by thesamet** [comment]
That actually makes total sense. Plugins are sandboxed and scalapb-validate's sandbox would have ScalaPB 0.10.10-preview6. I'll have a release later today or tomorrow that bring some syntax change as well (see scalapb/ScalaPB#1007), so it can be tested all at once.

**Comment by thesamet** [comment]
preview8 for both ScalaPB and ScalaPB-validate has been released. This change introduces a change to the field tranformation syntax, with documentation updated at https://github.com/scalapb/ScalaPB/commit/8d775ce1482755b00762c8d0574d239e214234a18d775ce1.

Summary of changes:
* Field transformations DSL is now part of ScalaPB. Therefore, the `field_transformations` list now goes into `scalapb.options`, either as file scoped or package scoped.
* Field transformations can match on any field extension options (not just PGV), so the syntax is more generic:
  ```protobuf
  field_transformations : [ 
    {when: { [validate.rules] { int32: {gt: 0}}}
     set:  { type: "scalapb.transforms.PositiveInt"}},

    {when: { [validate.rules] { int32 : {gt : 1}}}
      set: { type : "Int Refined Greater[W.`$([validate.rules].int32.gt)`.T]"}
      match_type : PRESENCE}
  ]
```
   Note the `[validate.rules]` reference to the PGV extension both in `when` and in path interpolations.
* ScalaPB-validate preprocessor role has been simplified to [this](https://scalapb.github.io/docs/validation#what-does-the-preprocessor-do). 
* Since field_transformations are applied by ScalaPB, the effect of opting out from a preprocessor via `-$name` as described in an [earlier comment](https://github.com/scalapb/scalapb-validate/issues/38#issuecomment-759132904) is different than before. Instead, it is possible now to opt out from field transformations altogether via a new ScalaPB file-level option (`ignore_field_transformations`).
* scalapb-validate-demo project has been updated with the new changes.

**Comment by bjaglin** [comment]
@thesamet Amazing work!

I tried out preview8 and found a regression related to the built-in transformations injected by the scalapb-validate preprocessor.

Here is a repro https://github.com/bjaglin/scalapb-validate-demo/commit/55cae34bbfe1f676b9d33458a722970442b4d20d
```
sbt:scalapb-validate-demo> protocGenerate 
[info] Compiling 2 protobuf files to /home/bjaglin/git/projects/scalapb-validate-demo/target/scala-2.13/src_managed/main/scalapb
[info] Compiling schema /home/bjaglin/git/projects/scalapb-validate-demo/src/main/protobuf/demo.proto
[info] Compiling schema /home/bjaglin/git/projects/scalapb-validate-demo/src/main/protobuf/options.proto
--jvm_1_out: options.proto: Could not find extension number 1071 for message 1071: {
  18: {scalapb-validate-demo / Compile / protocGenerate 1s
    1: 1
  }
}
```

I had a quick look, and it does look like ScalaPB currently needs to have a known mapping for field options used in transformations in the implementation (which it has incidentally on all tests/demos as custom transformations are defined), but I believe it's not a requirement as it could operate on unknown ids?

**Comment by thesamet** [comment]
The implementations attempts to resolve unknown ids by looking for extensions defined in the same file or the file that it directly imports. Removing the import of `validate/validate.proto` is the cause here. Since scalapb/validate.proto imports validate/validate.proto, then this wouldn't have happened if the search was recursive. I think it would be the right thing to do to eliminate this pitfall.

**Comment by bjaglin** [comment]
>  I believe it's not a requirement as it could operate on unknown ids?

I realized after writing this that we cannot operate on unknown ids if we want to be able [capture arbitrary option values with a human-readable path syntax `$()`](https://github.com/scalapb/scalapb-validate/issues/38#issuecomment-758488191).

> Since scalapb/validate.proto imports validate/validate.proto, then this wouldn't have happened if the search was recursive

I just tried your recursive fix, and unfortunately it's not enough as [the same problem happens for regular (non-package.proto) files which include neither `scalapb/validate.proto` nor `validate/validate.proto`](https://github.com/bjaglin/scalapb-validate/commit/4ad6846fb70321e35635a37fe1c09bff6416131e). I believe updating https://github.com/scalapb/ScalaPB/blob/76642ff0306c5c6ecfe944ba749c366846b8f2a8/compiler-plugin/src/main/scala/scalapb/compiler/FileOptionsCache.scala#L140 to accumulate extensions as we merge field transformations should be enough in that particular case.

**Comment by thesamet** [comment]
In preview11, `unique_to_set` and `cats_transforms` have been moved to `FileValidationOptions`, which is an extension of ScalaPB validation option. Previously, they were in their own file-level options. The preprocessor injects the field transformations when encountering those settings. The field transformations can then be propagated to the rest the package when `scope: PACKAGE` is set.  Since extensions in field transformations are resolved before propagation to other files, the situation described in the previous comment is no longer possible. Before preview11, field transformations were injected directly to each file under a package that had `(scalapb.validate.package).cats_transforms` which led to failure in extension resolution in files that didn't have the import.

**Comment by bjaglin** [comment]
Happy to report that with a local build of ScalaPB/scalapb-validate (preview11 + [my last fix](https://github.com/scalapb/ScalaPB/pull/1038)), we are utilizing all the features made available for this in our main project:
- built-in Cats transforms
- static, `EXACT`, adhoc transform based on PGV option
- dynamic, `PRESENCE`, adhoc transform based on custom options extracting values with path expansion
- package-level transformations opt-out (using package hierarchy)
- file-level transformations opt-out

From our side, that looks like the preview branch is ready to be merged!

**Comment by thesamet** [comment]
Released preview14. This release changes the DSL syntax again. The changes are:
* `FieldTransformation.when` matches on `google.protobuf.FieldDescriptorProto` (previously FieldOptions), so now it can match on things such as `type: TYPE_INT32` and `label: LABEL_REPEATED`
* `FieldTransformation.set` takes `google.protobuf.FieldOptions`, but the only field you allowed to send in this release is `[scalapb.field]` 
* Path value interpolation (`$(p)`) is now relative to the FieldDescriptorProto.
* Improvement in error messages.
* new syntax is updated in [docs](https://scalapb.github.io/docs/validation#rule-based-type-customization) via a61d421
* scalapb-validate-demo bumped up to preview14 and the new DSL syntax.

**Comment by bjaglin** [comment]
We have successfully integrated preview14 and started leveraging the more generic syntax to expanding the proto `type` on top of field options.

**Comment by thesamet** [comment]
> We have successfully integrated preview14 and started leveraging the more generic syntax to expanding the proto type on top of field options.

Fantastic! For ScalaPB, I merged the `0.10.x-preview` into `0.10.x` and released version `0.10.10`. The same changes have been forward-ported into the `0.11.x-preview` branch which got merged today into the master branch. Similarly, in ScalaPB-validate `0.2.x-preview` branch has been merged into master, and I released version `0.2.0`. The instructions in the [docs](https://scalapb.github.io/docs/validation) have been updated accordingly.
